Check and redirect to search query from 404 page, if results exist

suprith-hub opened the pull request:
@jcscottiii Could you please review this?

I have two main points to discuss:

**1. Handling Bad IDs**:
  Currently, I'm making an API call to fetch a single feature with the bad ID.
  Based on the response, I either show "Go to home" or "Search similar names" options.
  Other implementations I've seen seem to directly redirect to the homepage instead of a dedicated 404 page.
  Do you think this approach is better, or should we handle it differently?

**2. Placement of `handleNotFound` function**:
  I've placed `handleNotFound` in [mention file name].
Is this the right place, or would it be better suited elsewhere?
This fix addressing the issue: https://github.com/GoogleChrome/webstatus.dev/issues/1214

jcscottiii commented:
I will try to review this by the end of the day. You will probably need to update or add playwright tests to cover both the badID provided and not provided cases. And will want to add some unit tests too since this is now adding logic to the feature not found page.

Could you also chime in on https://github.com/GoogleChrome/webstatus.dev/issues/1263 with your specs and settings?

suprith-hub commented:
> I will try to review this by the end of the day. You will probably need to update or add playwright tests to cover both the badID provided and not provided cases. And will want to add some unit tests too since this is now adding logic to the feature not found page.
> 
> Could you also chime in on #1263 with your specs and settings?

@jcscottiii I will add playwright test for `Search for similar names`.
I have some doubts regarding GSoC project: Search UI enhancements.

To implement search in webstatus and chrome-dashboard, can I use RediSearch or is ElasticSearch better.
I am planning to research on RediSearch.
Am I in right direction, am I supposed to use these tools ?
Where can I ask doubts like these

jcscottiii commented:
@suprith-hub 
> GSoC project: Search UI enhancements

The work for that proposal will happen primarily in https://github.com/GoogleChrome/chromium-dashboard then later be ported to here.

You should create an issue in https://github.com/GoogleChrome/chromium-dashboard

As a heads up, we try to use managed services in GCP. This reduces the maintenance burden. So if you're looking into RediSearch, check if Memorystore for Redis supports it ([docs](https://cloud.google.com/memorystore/docs/redis)). Plus, the chromium-dashboard already uses redis so it simplifies the changes needed to the infrastructure.

I don't think there's a managed version of ElasticSearch in GCP. But feel free to see if there are other GCP managed services that could help.

These are all just some of my thoughts. But you should create the issue in that repo to start the discussion before diving too deep.

cc: @jrobbins for context

jcscottiii reviewed (commented):
@suprith-hub For this change, I see that you are calling getFeatures to check if there are any features with similar names. 

All that said, I would prefer to show suggestions like mentioned in #1214 but I'm not sure how much time you have to commit to this. But I appreciate any and all contributions. If you don't have much time, just show the second button to search for similar features.

---

If we are going to call getFeatures, we should present some of the options. Otherwise, we could just add the second button alongside the existing go home button.

Also, you should use Lit's Task for making asynchronous network calls like that instead of using the apiclient directly. More about it here: https://lit.dev/docs/data/task/

Also, I would move the API client call itself to the notfound error page.



Other notes:

We should move the creation of the 404 feature page URL string to https://github.com/GoogleChrome/webstatus.dev/blob/main/frontend/src/static/js/utils/urls.ts

Also inside that URL.ts helper, you should do the retrieval of the feature ID instead of pulling from the `window` object directly.

suprith-hub commented:
> @suprith-hub For this change, I see that you are calling getFeatures to check if there are any features with similar names.
> 
> All that said, I would prefer to show suggestions like mentioned in #1214 but I'm not sure how much time you have to commit to this. But I appreciate any and all contributions. If you don't have much time, just show the second button to search for similar features.
> 
> If we are going to call getFeatures, we should present some of the options. Otherwise, we could just add the second button alongside the existing go home button.
> 
> Also, you should use Lit's Task for making asynchronous network calls like that instead of using the apiclient directly. More about it here: https://lit.dev/docs/data/task/
> 
> Also, I would move the API client call itself to the notfound error page.
> 
> Other notes:
> 
> We should move the creation of the 404 feature page URL string to https://github.com/GoogleChrome/webstatus.dev/blob/main/frontend/src/static/js/utils/urls.ts
> 
> Also inside that URL.ts helper, you should do the retrieval of the feature ID instead of pulling from the `window` object directly.

@jcscottiii about showing similar results,
if there are more results then there has to be pagiination. So, should I make a UI which will just have names with pagination ?
If yes: What will be the page size(should it be dynamic)

jcscottiii commented:
It is not meant to replace the overview page. But only include suggestions. It does not need to show all of the possible results like the overview page.

Instead, it could:
- Show up to 5 results. The phrasing could be like:
```
We could not find Feature ID: <featureID>

Here are some similar features:
- A
- B
- C
- D
- E

Button to Search for more similar features | Button to return home | Button to report bugs
```

The list and the first button would only appear if it found features.

Hopefully that makes sense. Let me know if you have any questions.

suprith-hub commented:
> It is not meant to replace the overview page. But only include suggestions. It does not need to show all of the possible results like the overview page.
> 
> Instead, it could:
> 
> * Show up to 5 results. The phrasing could be like:
> 
> ```
> We could not find Feature ID: <featureID>
> 
> Here are some similar features:
> - A
> - B
> - C
> - D
> - E
> 
> Button to Search for more similar features | Button to return home | Button to report bugs
> ```
> 
> The list and the first button would only appear if it found features.
> 
> Hopefully that makes sense. Let me know if you have any questions.

@jcscottiii I will make the changes by tomorrow, and
I have raised an issue regarding GSoC Search tool suggestion. I previously sent an email asjing doubts regarding the project, but I’m not sure if it was received. Could there be any domain restrictions on emails? If there’s a preferred way to contact you, please let me know.

suprith-hub commented:
![Screenshot from 2025-03-22 00-47-57](https://github.com/user-attachments/assets/8f990636-c915-48c4-b9bc-7f7c93fc63f6)
![Screenshot from 2025-03-22 00-48-09](https://github.com/user-attachments/assets/a993ddf6-2833-450c-910a-11b2ba169302)
![Screenshot from 2025-03-22 00-48-26](https://github.com/user-attachments/assets/b36bd8ab-2e4e-4b48-98b8-3703e09c9db3)
@jcscottiii does these look good ?
Can you suggest if changes are necessary...
I will make a draft commit which I will revert after suggestions

jcscottiii commented:
@suprith-hub Thank you for the updates with the three screenshots/cases. We should ignore the second case for now. If there are no suggestions, there's no need to show the Search for similar features button.

suprith-hub commented:
> @suprith-hub Thank you for the updates with the three screenshots/cases. We should ignore the second case for now. If there are no suggestions, there's no need to show the Search for similar features button.

@jcscottiii I have put tests and made latest push, you can review it completely

jcscottiii commented:
@suprith-hub I have not forgotten about this. I will get around to reviewing this tomorrow.

suprith-hub commented:
> @suprith-hub I have not forgotten about this. I will get around to reviewing this tomorrow.

Yeah, no worries at all! I figured you might’ve scheduled it — thanks for letting me know! 😊

jcscottiii commented on frontend/src/static/js/components/webstatus-notfound-error-page.ts:
```diff
@@ -50,75 +98,137 @@ export class WebstatusNotFoundErrorPage extends LitElement {
   protected render(): TemplateResult {
+    const featureId = getSearchQuery(this.location);
+
+    console.log('hello', featureId);
```
```suggestion
    console.log('hello', featureId);
```
Remove this

suprith-hub replied:
Sorry, Iis my mistake..

jcscottiii commented on frontend/src/static/js/components/webstatus-notfound-error-page.ts:
```diff
@@ -15,12 +15,61 @@
+              }))
+            : [];
+        } catch (fetchError) {
+          console.error('Error fetching similar features:', fetchError);
```
we should use the Shoelace toast mechanism instead of printing to console. Example: https://github.com/GoogleChrome/webstatus.dev/blob/3764c2acd63fcb32b9d1cd347f49ee3ba18d0a2d/frontend/src/static/js/components/webstatus-overview-filters.ts#L458-L462

jcscottiii commented on e2e/tests/404.spec.ts:
```diff
@@ -35,12 +37,134 @@ test('Bad URL redirection to 404 page', async ({page}) => {
+          data.data.length,
+        );
+
+        // ✅ Click first similar feature
```
```suggestion
        // Click first similar feature
```

Could you remove the emojis from here and other comments?

jcscottiii commented on e2e/tests/404.spec.ts:
```diff
@@ -35,12 +37,134 @@ test('Bad URL redirection to 404 page', async ({page}) => {
+    await test.step(`Testing API response for: ${badUrl}`, async () => {
+      await page.goto(badUrl);
+      await expect(page).toHaveURL(
+        'http://localhost:5555/errors-404/feature-not-found?q=' + query,
```
```suggestion
        `${BASE_URL}/errors-404/feature-not-found?q=${query}`
```

jcscottiii commented on e2e/tests/404.spec.ts:
```diff
@@ -35,12 +37,134 @@ test('Bad URL redirection to 404 page', async ({page}) => {
+  page,
+}) => {
+  // Test URLs for different feature IDs
+  const badUrls = [
```
There seems to be overlap between this test which loops through two urls and the `matches the screenshot with similar results` test (which also searches for query = q).

This test has a lot of great validation steps that should instead be made into helper methods in this test file

We should have two tests:
- Should show the home and report buttons when there are no similar features
- Should show the similar features, home and report buttons when there are similar features

In each test, you could use those common helpers. As well as you won't have to the e2e test make an out of band request to find out if there are similar features or not. 

Take a look at https://github.com/GoogleChrome/webstatus.dev/blob/main/e2e/tests/overview-page.spec.ts for how we use some helper methods across multiple tests.

suprith-hub replied:
@jcscottiii I have moved the functions to utils file.

jcscottiii commented on e2e/tests/404.spec.ts:
```diff
@@ -35,12 +37,134 @@ test('Bad URL redirection to 404 page', async ({page}) => {
+      // Assert that the response status code is 404
+      expect(featurePageResponse.status()).toBe(404);
+
+      // Mock API response for similar features
+      const apiUrl = API_BASE_URL.replace('{query}', query);
+      const response = await page.context().request.get(apiUrl);
+      expect(response.status()).toBe(200);
+
+      const data = await response.json();
+      const hasResults = Array.isArray(data?.data) && data.data.length > 0;
```
We should not make an out of band request to figure out if the query has results. We should know beforehand. See my above comment

suprith-hub replied:
Should I replace it with screen shot instead? I will make commit with that change.

jcscottiii commented on frontend/src/static/js/components/webstatus-notfound-error-page.ts:
```diff
@@ -50,75 +98,137 @@ export class WebstatusNotFoundErrorPage extends LitElement {
-          <div id="error-action-report">
+        ${featureId
+          ? html`
+              ${this._loadingSimilarResults?.status === TaskStatus.PENDING
```
Instead of comparing the status at render time, we should instead declare what we want to happen in the task's render method

https://lit.dev/docs/data/task/#rendering-tasks

suprith-hub replied:
I have used lit tasks for rendering these states,..

jcscottiii commented on frontend/src/static/js/components/webstatus-notfound-error-page.ts:
```diff
@@ -50,75 +98,137 @@ export class WebstatusNotFoundErrorPage extends LitElement {
-      <div id="error-container">
+      <div
+        id="error-container"
+        style="gap: ${this.similarFeatures.length > 0 ? '32px' : '48px'};"
```
We should not inline styles. Instead have a different class for each case. and assign the class appropriately

jcscottiii commented on frontend/src/static/js/components/webstatus-notfound-error-page.ts:
```diff
@@ -50,75 +98,137 @@ export class WebstatusNotFoundErrorPage extends LitElement {
+        <div
+          id="error-actions"
+          style="gap: ${featureId && this.similarFeatures.length > 0
+            ? '16px'
```
We should not inline styles. Instead have a different class for each case. and assign the class appropriately

jcscottiii commented on frontend/src/static/js/components/webstatus-notfound-error-page.ts:
```diff
@@ -50,75 +98,137 @@ export class WebstatusNotFoundErrorPage extends LitElement {
+                </sl-button>
+              `
+            : ''}
+          <sl-button id="error-action-home-btn" variant="default" href="/">
```
If there are no similar features, this button should have the primary class.

jcscottiii reviewed (changes requested):
Thanks for the contribution. Take a look at these comments and let me know if you have any questions.

suprith-hub commented on frontend/src/static/js/components/webstatus-notfound-error-page.ts:
```diff
@@ -50,75 +102,152 @@ export class WebstatusNotFoundErrorPage extends LitElement {
   protected render(): TemplateResult {
+    const featureId = getSearchQuery(this.location);
+    const taskState = this._similarResults?.value;
+    const hasSimilar = Array.isArray(taskState) && taskState.length > 0;
```
@jcscottiii Can I declares variables to check statuses, here... is it a good way?

suprith-hub commented:
@jcscottiii I have made the changes. 
Merged snapshot test and removed API call in this test `shows similar features and all buttons when results exist`.
Use lit Task.render() instead of property similarResults.
I have 2 doubts:
1. Is is correct to define variables for classNames styling.
2. Can I use classnames like gap-16, gap-32 ?

suprith-hub commented:
@jcscottiii I have rebased the branch and made latest changes, I think the PR can be reviewed now.

jcscottiii commented:
FYI: I was away last week. Will be reviewing this today.

suprith-hub commented:
> FYI: I was away last week. Will be reviewing this today.

Yeah, ok @jcscottiii.
Also can I pick up other issues, like pending PR of github username...

jcscottiii commented on frontend/src/static/js/components/webstatus-notfound-error-page.ts:
```diff
@@ -50,75 +102,152 @@ export class WebstatusNotFoundErrorPage extends LitElement {
+          text-decoration: underline;
+          color: #0056b3;
+        }
+        .gap-16 {
```
For the sake of simplicity, we can probably get away with just having one gap.


Instead, use `gap: var(--content-padding);` where needed.

suprith-hub replied:
Ok we will use single gap-16 ...

suprith-hub commented on frontend/src/static/js/components/webstatus-notfound-error-page.ts:
```diff
@@ -50,75 +102,152 @@ export class WebstatusNotFoundErrorPage extends LitElement {
     ];
   }
+
   protected render(): TemplateResult {
```
Wow! nice, this makes the code much readable.
I will definitely follow this, thank you!

suprith-hub commented on frontend/src/static/js/components/webstatus-notfound-error-page.ts:
```diff
@@ -50,75 +102,152 @@ export class WebstatusNotFoundErrorPage extends LitElement {
+
   protected render(): TemplateResult {
+    const featureId = getSearchQuery(this.location);
+    const taskState = this._similarResults?.value;
```
I will remove these variables and `taskState` since we are using single `gap-16`

suprith-hub commented:
@jcscottiii  I have refactored, the code looks pretty good now(readability wise).
I have removed complexity in gap spacing with its corresponding unwanted variables

jcscottiii commented:
@suprith-hub Please run `make lint-fix`. Also, feel free to run the `make precommit` command to run the same command used in the build stage of CI.

jcscottiii commented on frontend/src/static/js/components/test/webstatus-notfound-error-page.test.ts:
```diff
@@ -0,0 +1,176 @@
+ */
+
+import {expect, fixture, html} from '@open-wc/testing';
+import '../webstatus-not-found-error-page.js';
```
```suggestion
import '../webstatus-notfound-error-page.js';
```

This is the typo that you have causing the test file to not run. After you fix it, you will need to look at the failing tests.

suprith-hub replied:
Hey @jcscottiii — I was away due to some eye strain.

The test failures were due to a misunderstanding on my part — I had assumed unit tests would call the real API, but they required mocking. I’ve now updated the component with clearer naming and refactored the tests to use a reusable mock API function.

Precommit checks are all passing now. I believe it’s ready for review!

Let me know if anything else is needed 🙌
